pkg/operator/cluster/ceph/osd/osd.go

```diff
@@ -105,260 +140,443 @@ func (c *Cluster) Start() error {
 		} else {
 			logger.Infof("osd daemon set started")
 		}
-	} else {
-		for i := range c.Storage.Nodes {
-			// fully resolve the storage config for this node
-			n := c.Storage.ResolveNode(c.Storage.Nodes[i].Name)
 
-			resources := k8sutil.MergeResourceRequirements(c.Storage.Nodes[i].Resources, c.resources)
+		return nil
+	}
 
-			// create the replicaSet that will run the OSDs for this node
-			rs := c.makeReplicaSet(n.Name, n.Devices, n.Selection, resources, n.Config)
-			_, err := c.context.Clientset.Extensions().ReplicaSets(c.Namespace).Create(rs)
-			if err != nil {
-				if !errors.IsAlreadyExists(err) {
-					return fmt.Errorf("failed to create osd replica set for node %s. %+v", n.Name, err)
-				}
-				logger.Infof("osd replica set already exists for node %s", n.Name)
+	// orchestrate individual nodes, starting with any that are still ongoing (in the case that we
+	// are resuming a previous orchestration attempt)
+	if inProgressNode, status := c.findInProgressNode(); inProgressNode != "" {
+		logger.Infof("resuming orchestration of in progress node %s, status: %+v", inProgressNode, status)
+		if err := c.waitForCompletion(inProgressNode); err != nil {
+			logger.Warningf("failed waiting for in progress node %s, will continue with orchestration.  %+v", inProgressNode, err)
+		}
+	}
+
+	errorMessages := make([]string, 0)
+
+	// start with nodes currently in the storage spec
+	for i := range c.Storage.Nodes {
```

Comment:
this is synchronous for the operator, right? All cluster CRD processing will be blocked until this potentially very long running operation is completed? For example, nobody can delete the crd or create another one until this is done? I wonder if we can make this asynchronous so that other operations can still be done.